proposals/0000-import-shadowing.rst

```diff
@@ -0,0 +1,311 @@
+Import shadowing
+================
+
+.. author:: Gergo Erdi
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/652>`_.
+.. sectnum::
+.. contents::
+
+Local binders are allowed to shadow names defined in outer
+scopes. Here, we propose that top-level binders should also be allowed
+to shadow imported names, basically treating imported modules as
+living in a scope that is "outside" the top-level one.
+
+
+Motivation
+----------
+Currently, there is no way in Haskell to shadow an imported name. The
+only workaround is to avoid importing said name at all. For example,
+if we want to use some ``Control.Exception`` functions while also
+defining our own ``catch``, we have to import the former with an explicit
+hiding of ``catch``:
+
+::
+
+ module Example1 where
+
+ import Control.Exception hiding (catch)
+
+ catch :: Fish -> IO Food
+ catch fish = ...
+
+ goFishing :: IO Food
+ goFishing = bracket goToCabin goHome $ catch salmon
+
+Without hiding ``catch``, this results in a name resolution conflict
+in the definition of ``goFishing``:
+
+::
+
+ Ambiguous occurrence `catch'
+ It could refer to
+    either `Control.Exception.catch',
+           imported from `Control.Exception' at Example1.hs:1:1-24
+           (and originally defined in `GHC.IO')
+        or `Example1.catch', defined at Example1.hs:7:1
+
+Our proposal is an alternative name resolution policy exposed as a
+language extension tentatively named ``ImportShadowing``. It allows
+definitions from the current module's top-level scope to shadow names
+from imported modules. So in this example, the occurrence of ``catch``
+in the definition of ``goFishing`` resolves to the definition in
+``Example1`` even when ``catch`` is not explicitly hidden from the
+import list of ``Control.Exception``.
+
+We believe this is an improvement to the status quo because it is a
+natural extension of the idea behind the shadowing policy of local
+binders: that shadowing expresses the intention by the user to only
+care about the names that are defined "more nearby".
+
+Proposed Change Specification
```

Comment:
This section does not read as a _specification_ to me. Instead, I read it as an interpretation of an unwritten specification by means of desugaring, and that desugaring is itself specified only by example.

Happily, I think the specification is very easy:

* Occurrences of an identifier in the code of a module and in its export list are looked up in an environment containing all the global definitions in that module. If that lookup fails, then the identifier is looked up in an environment containing all the definitions imported by `import` statements (including the implicit `import` of the `Prelude`, if any).

I think that covers it!